haystack/components/builders/hypothetical_doc_query.py

```diff
@@ -0,0 +1,127 @@
+from haystack import Pipeline, component, Document
+from haystack.components.embedders.sentence_transformers_document_embedder import SentenceTransformersDocumentEmbedder
+from haystack.components.generators.openai import OpenAIGenerator
+from haystack.components.builders import PromptBuilder
+
+from typing import List, Dict, Any
+from numpy import array, mean
+
+
+@component
+class InstructLM:
+    """
+    Subcomponent to generate hypothetical documents using an instruction-following language model (InstructLM).
+    """
+
+    def __init__(self, instruct_llm, nr_completions):
+        self.prompt_builder = PromptBuilder(
+            template="""
+        Given a question, generate a paragraph of text that answers the question.
+        Question: {{question}}
+        Paragraph:
+        """
+        )
+        self.generator = OpenAIGenerator(
+            model=instruct_llm, generation_kwargs={"n": nr_completions, "temperature": 0.75, "max_tokens": 400}
+        )
+
+        self.pipeline = Pipeline()
+        self.pipeline.add_component(name="prompt_builder", instance=self.prompt_builder)
+        self.pipeline.add_component(name="generator", instance=self.generator)
+        self.pipeline.connect("prompt_builder", "generator")
+
+    def to_dict(self) -> Dict[str, Any]:
+        # ToDo
+        pass
+
+    def from_dict(cls, data: Dict[str, Any]) -> "InstructLM":
+        # ToDo
+        pass
+
+    @component.output_types(replies=List[str])
+    def run(self, query: str):
+        result = self.pipeline.run(data={"prompt_builder": {"question": query}})
+        return result["generator"]
+
+
+@component
+class HyDEBuilder:
```

Comment:
I would give that component a more generic name so it can be reused or better keep it as two separate components. One document embedder and one new average pooling component.